engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoSource.java

```diff
@@ -1,44 +1,72 @@
 package org.enso.interpreter.runtime.data;
 
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
 import com.oracle.truffle.api.source.Source;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.data.text.Text;
+import org.enso.interpreter.runtime.data.vector.ArrayLikeHelpers;
 
 /** Wrapper for exposing sources to Enso. Delegates to original methods with no behavior changes. */
-public class EnsoSource {
+@ExportLibrary(InteropLibrary.class)
+public final class EnsoSource implements EnsoObject {
+  private static final String[] MEMBERS = {
```

Comment:
Why do we change to this 'switch-based' way of invoking methods on these helper classes?

The old way seemed much more readable, but I assume the change is needed for some reason?

Reply:
Yes, it is important. It **avoids reflection**. The [InteropLibrary](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/InteropLibrary.html) is a way to do _reflection in Truffle languages_ in a away more effective way compared to `java.lang.reflect` approach.

Neither `EnsoSource` or `EnsoSourceSection` should have been using JVM reflection to begin with. It just remained unnoticed until we needed to use them from `native-image` built executable.

The appropriate commit is e9551b7e621b05d880dfd61ec7b647a6a60e5119

Reply:
Thanks for the explanation